File: backend/domain/workflow/entity/vo/node_test.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package vo

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestTypeInfoToJSONSchema(t *testing.T) {
	tests := []struct {
		name     string
		typeInfo map[string]*TypeInfo
		validate func(t *testing.T, schema string)
	}{
		{
			name: "Basic Data Types",
			typeInfo: map[string]*TypeInfo{
				"stringField": {Type: DataTypeString},
				"intField":    {Type: DataTypeInteger},
				"numField":    {Type: DataTypeNumber},
				"boolField":   {Type: DataTypeBoolean},
				"timeField":   {Type: DataTypeTime},
			},
			validate: func(t *testing.T, schema string) {
				var schemaObj map[string]any
				err := json.Unmarshal([]byte(schema), &schemaObj)
				assert.NoError(t, err)

				props := schemaObj["properties"].(map[string]any)

				// Validate string field
				stringProp := props["stringField"].(map[string]any)
				assert.Equal(t, "string", stringProp["type"])

				// Validate integer fields
				intProp := props["intField"].(map[string]any)
				assert.Equal(t, "integer", intProp["type"])

				// Validate numeric field
				numProp := props["numField"].(map[string]any)
				assert.Equal(t, "number", numProp["type"])

				// Validate Boolean fields
				boolProp := props["boolField"].(map[string]any)
				assert.Equal(t, "boolean", boolProp["type"])

				// validation time field
				timeProp := props["timeField"].(map[string]any)
				assert.Equal(t, "string", timeProp["type"])
				assert.Equal(t, "date-time", timeProp["format"])
			},
		},
		{
			name: "Complex Data Types",
			typeInfo: map[string]*TypeInfo{
				"objectField": {Type: DataTypeObject},
				"arrayField": {
					Type:         DataTypeArray,
					ElemTypeInfo: &TypeInfo{Type: DataTypeString},
				},
				"fileField": {
					Type:     DataTypeFile,
					FileType: fileSubTypePtr(FileTypeImage),
				},
			},
			validate: func(t *testing.T, schema string) {
				var schemaObj map[string]any
				err := json.Unmarshal([]byte(schema), &schemaObj)
				assert.NoError(t, err)

				props := schemaObj["properties"].(map[string]any)

				// Validation Object Field
				objProp := props["objectField"].(map[string]any)
				assert.Equal(t, "object", objProp["type"])

				// Validate array fields
				arrProp := props["arrayField"].(map[string]any)
				assert.Equal(t, "array", arrProp["type"])
				items := arrProp["items"].(map[string]any)
				assert.Equal(t, "string", items["type"])

				// Validate file field
				fileProp := props["fileField"].(map[string]any)
				assert.Equal(t, "string", fileProp["type"])
				assert.Equal(t, "image", fileProp["contentMediaType"])
			},
		},
		{
			name: "Nested Array",
			typeInfo: map[string]*TypeInfo{
				"nestedArray": {
					Type:         DataTypeArray,
					ElemTypeInfo: &TypeInfo{Type: DataTypeObject},
				},
			},
			validate: func(t *testing.T, schema string) {
				var schemaObj map[string]any
				err := json.Unmarshal([]byte(schema), &schemaObj)
				assert.NoError(t, err)

				props := schemaObj["properties"].(map[string]any)

				// Validate nested array fields
				arrProp := props["nestedArray"].(map[string]any)
				assert.Equal(t, "array", arrProp["type"])
				items := arrProp["items"].(map[string]any)
				assert.Equal(t, "object", items["type"])
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			schema, err := TypeInfoToJSONSchema(tt.typeInfo, nil)
			assert.NoError(t, err)
			tt.validate(t, schema)
		})
	}
}

// Helper functions for creating DataType pointers
func stringPtr(dt DataType) *DataType {
	return &dt
}

// Helper function for creating a FileSubType pointer
func fileSubTypePtr(fst FileSubType) *FileSubType {
	return &fst
}

File: frontend/packages/workflow/base/__tests__/types/node.test.ts
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, it, expect } from 'vitest';

import { StandardNodeType } from '../../src/types/node-type';
import type { WorkflowNodeJSON, WorkflowJSON } from '../../src/types/node';

describe('node types', () => {
  describe('WorkflowNodeJSON', () => {
    it('应该能够创建基本的节点 JSON', () => {
      const node: WorkflowNodeJSON = {
        id: '1',
        type: StandardNodeType.Start,
        data: {},
      };

      expect(node).toEqual({
        id: '1',
        type: StandardNodeType.Start,
        data: {},
      });
    });

    it('应该能够创建带有元数据的节点 JSON', () => {
      const node: WorkflowNodeJSON = {
        id: '1',
        type: StandardNodeType.Start,
        meta: {
          position: { x: 100, y: 100 },
        },
        data: {},
      };

      expect(node.meta).toBeDefined();
      expect(node.meta?.position).toEqual({ x: 100, y: 100 });
    });

    it('应该能够创建带有版本的节点 JSON', () => {
      const node: WorkflowNodeJSON = {
        id: '1',
        type: StandardNodeType.Start,
        data: {},
        version: '1.0.0',
      };

      expect(node.version).toBe('1.0.0');
    });

    it('应该能够创建带有子节点的节点 JSON', () => {
      const node: WorkflowNodeJSON = {
        id: '1',
        type: StandardNodeType.Start,
        data: {},
        blocks: [
          {
            id: '2',
            type: StandardNodeType.Code,
            data: {},
          },
        ],
      };

      expect(node.blocks).toHaveLength(1);
      expect(node.blocks?.[0].type).toBe(StandardNodeType.Code);
    });

    it('应该能够创建带有边的节点 JSON', () => {
      const node: WorkflowNodeJSON = {
        id: '1',
        type: StandardNodeType.Start,
        data: {},
        edges: [
          {
            sourceNodeID: '1',
            targetNodeID: '2',
          },
        ],
      };

      expect(node.edges).toHaveLength(1);
      expect(node.edges?.[0].sourceNodeID).toBe('1');
      expect(node.edges?.[0].targetNodeID).toBe('2');
    });

    it('应该能够使用泛型数据类型', () => {
      interface CustomData {
        name: string;
        value: number;
      }

      const node: WorkflowNodeJSON<CustomData> = {
        id: '1',
        type: StandardNodeType.Start,
        data: {
          name: 'test',
          value: 42,
        },
      };

      expect(node.data.name).toBe('test');
      expect(node.data.value).toBe(42);
    });
  });

  describe('WorkflowJSON', () => {
    it('应该能够创建基本的工作流 JSON', () => {
      const workflow: WorkflowJSON = {
        nodes: [],
        edges: [],
      };

      expect(workflow.nodes).toEqual([]);
      expect(workflow.edges).toEqual([]);
    });

    it('应该能够创建包含节点和边的工作流 JSON', () => {
      const workflow: WorkflowJSON = {
        nodes: [
          {
            id: '1',
            type: StandardNodeType.Start,
            data: {},
          },
          {
            id: '2',
            type: StandardNodeType.End,
            data: {},
          },
        ],
        edges: [
          {
            sourceNodeID: '1',
            targetNodeID: '2',
          },
        ],
      };

      expect(workflow.nodes).toHaveLength(2);
      expect(workflow.edges).toHaveLength(1);
      expect(workflow.nodes[0].type).toBe(StandardNodeType.Start);
      expect(workflow.nodes[1].type).toBe(StandardNodeType.End);
      expect(workflow.edges[0].sourceNodeID).toBe('1');
      expect(workflow.edges[0].targetNodeID).toBe('2');
    });
  });
});
